core/lib/zksync_core/src/reorg_detector/mod.rs

```diff
@@ -361,127 +393,31 @@ impl ReorgDetector {
             known_valid_l1_batch.0,
             diverged_l1_batch.0,
             |number| async move {
-                Ok(match self.root_hashes_match(L1BatchNumber(number)).await? {
-                    MatchOutput::Match | MatchOutput::NoRemoteReference => true,
-                    MatchOutput::Mismatch => false,
-                })
+                match self.root_hashes_match(L1BatchNumber(number)).await {
+                    Err(HashMatchError::RemoteHashMissing) => Ok(true),
+                    res => res,
+                }
             },
         )
         .await
         .map(L1BatchNumber)
     }
 
-    pub async fn run(
-        mut self,
-        mut stop_receiver: watch::Receiver<bool>,
-    ) -> anyhow::Result<Option<L1BatchNumber>> {
+    pub async fn run(mut self, stop_receiver: watch::Receiver<bool>) -> Result<(), Error> {
         self.event_handler.initialize();
-        loop {
-            match self.run_inner(&mut stop_receiver).await {
-                Ok(l1_batch_number) => return Ok(l1_batch_number),
-                Err(HashMatchError::Rpc(err)) if is_transient_err(&err) => {
-                    tracing::warn!("Following transport error occurred: {err}");
+        while !*stop_receiver.borrow() {
+            match self.check_consistency().await {
+                Err(err) if !err.is_transient() => return Err(err),
+                Err(err) => {
```

Comment:
Bikeshedding: IMO, it'd be more straightforward to place this clause first with `if err.is_transient()`; in this way, the condition wouldn't have to be inverted.